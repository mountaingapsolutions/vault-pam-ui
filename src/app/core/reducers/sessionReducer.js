import sessionAction from 'app/core/actions/sessionAction';

/**
 * Reducer class responsible for reacting to and handling the session REST calls.
 *
 */

/**
 * The primary reduce method.
 *
 * @public
 * @param {Object} previousState - The previous state.
 * @param {Object} action - Action type.
 * @returns {Object} The updated state.
 */
export default (previousState = {
    authUser: {},
    vaultDomain: {},
    vaultLookupSelf: {},
    sealStatus: {},
    vaultToken: {}
}, action) => {
    switch (action.type) {
        case sessionAction.ACTION_TYPES.AUTHENTICATE_USER_PASS:
            // Message the error message with proper grammar.
            if (action.errors) {
                action.errors = action.errors.map(err => `Authentication failed: ${err}.`);
            }
            return {...previousState, authUser: action};
        case sessionAction.ACTION_TYPES.SET_DOMAIN:
            return {...previousState, vaultDomain: action};
        case sessionAction.ACTION_TYPES.SET_TOKEN:
            return {...previousState, vaultToken: action};
        case sessionAction.ACTION_TYPES.VALIDATE_DOMAIN:
            return {...previousState, sealStatus: action};
        case sessionAction.ACTION_TYPES.VALIDATE_TOKEN:
            // Message the error message with proper grammar.
            if (action.errors) {
                action.errors = action.errors.map(err => `Authentication failed: ${err}.`);
            }
            return {...previousState, vaultLookupSelf: action};
        default:
            return {...previousState};
    }
};
