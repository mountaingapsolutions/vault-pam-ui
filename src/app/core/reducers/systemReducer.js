import systemAction from 'app/core/actions/systemAction';

/**
 * Reducer class responsible for reacting to and handling the system REST calls.
 *
 */

/**
 * The primary reduce method.
 *
 * @public
 * @param {Object} previousState - The previous state.
 * @param {Object} action - Action type.
 * @returns {Object} The updated state.
 */
export default (previousState = {
    config: {},
    selfCapabilities: {},
    sealStatus: {}
}, action) => {
    switch (action.type) {
        case systemAction.ACTION_TYPES.GET_CONFIG:
            return {...previousState, config: action.data || {}, action};
        case systemAction.ACTION_TYPES.GET_SELF_CAPABILITIES:
            return {...previousState, selfCapabilities: action.data || {}, action};
        case systemAction.ACTION_TYPES.GET_SEAL_STATUS:
            return {...previousState, sealStatus: action.data || {}, action};
        default:
            return {...previousState};
    }
};
