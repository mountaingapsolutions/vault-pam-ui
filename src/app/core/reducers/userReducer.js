import userAction from 'app/core/actions/userAction';

/**
 * Reducer class responsible for reacting to and handling user REST calls.
 *
 */

/**
 * The primary reduce method.
 *
 * @public
 * @param {Object} previousState - The previous state.
 * @param {Object} action - Action type.
 * @returns {Object} The updated state.
 */
export default (previousState = {
    user: {},
    isLoggedIn: true
}, action) => {
    switch (action.type) {
        case userAction.ACTION_TYPES.GET_USER:
            return {
                ...previousState,
                user: action.data
            };
        case userAction.ACTION_TYPES.LOGOUT:
            const isLoggedIn = action.data && action.data.status && action.data.status !== 'ok';
            return {...previousState, isLoggedIn};
        default:
            return {...previousState};
    }
};
