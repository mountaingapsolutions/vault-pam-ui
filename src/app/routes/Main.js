/* global window */

import {withStyles} from '@material-ui/core/styles';
import {
    AppBar,
    Badge,
    Card,
    CardContent,
    Grid,
    IconButton,
    List,
    ListItem,
    ListItemIcon,
    ListItemSecondaryAction,
    ListItemText,
    Menu,
    MenuItem,
    Snackbar,
    Toolbar,
    Typography
} from '@material-ui/core';
import AccountCircle from '@material-ui/icons/AccountCircle';
import Button from 'app/core/components/common/Button';
import CloseIcon from '@material-ui/icons/Close';
import LockIcon from '@material-ui/icons/Lock';
import KeyboardArrowRightIcon from '@material-ui/icons/KeyboardArrowRight';
import ListIcon from '@material-ui/icons/List';
import LockOpenIcon from '@material-ui/icons/LockOpen';
import NotificationsIcon from '@material-ui/icons/Notifications';
import PropTypes from 'prop-types';
import React, {Component} from 'react';
import {connect} from 'react-redux';
import {Link, Redirect, Route, Switch, withRouter} from 'react-router-dom';
import kvAction from 'app/core/actions/kvAction';
import sessionAction from 'app/core/actions/sessionAction';
import systemAction from 'app/core/actions/systemAction';
import userAction from 'app/core/actions/userAction';
import SecretRequestQueueModal from 'app/core/components/secretRequest/SecretRequestQueueModal';
import UserProfileModal from 'app/core/components/UserProfileModal';
import Footer from 'app/core/components/common/Footer';
import NotificationsModal from 'app/core/components/NotificationsModal';
import SecretsList from 'app/routes/secrets/SecretsList';
import Constants from 'app/util/Constants';
import localStorageUtil from 'app/util/localStorageUtil';

/**
 * The main container.
 */
class Main extends Component {

    /**
     * The constructor method. Executed upon class instantiation.
     *
     * @public
     * @param {Object} props Props to initialize with.
     */
    constructor(props) {
        super(props);

        this.state = {
            accountAnchorElement: null,
            isSecretRequestsModalOpen: false,
            isUserProfileModalOpen: false,
            notificationAnchorElement: null,
            showRootWarning: false
        };
        this._closeModal = this._closeModal.bind(this);
        this._onClose = this._onClose.bind(this);
        this._openModal = this._openModal.bind(this);
        this._toggleAccountMenu = this._toggleAccountMenu.bind(this);
    }

    /**
     * Handle for when a close event from the snackbar is triggered.
     *
     * @private
     * @param {SyntheticMouseEvent} event The event.
     * @param {string} reason The close reason.
     */
    _onClose(event, reason) {
        if (reason !== 'clickaway') {
            this.setState({
                showRootWarning: false
            });
        }
    }

    /**
     * Handle for when a close event from the modal is clicked.
     *
     * @private
     * @param {string} modalState The close reason.
     */
    _closeModal(modalState) {
        this.setState({[modalState]: false});
    }

    /**
     * Handle for when a modal is opened.
     *
     * @private
     * @param {string} modalState The close reason.
     */
    _openModal(modalState) {
        this.setState({[modalState]: true, accountAnchorElement: null});
    }

    /**
     * Toggles the account menu.
     *
     * @private
     * @param {SyntheticMouseEvent} event The event.
     */
    _toggleAccountMenu(event) {
        const {accountAnchorElement} = this.state;
        this.setState({
            accountAnchorElement: accountAnchorElement ? null : event.currentTarget
        });
    }

    /**
     * Returns the proper icon from the provided mount type.
     *
     * @private
     * @param {string} type The mount type.
     * @returns {Element}
     */
    _renderIconFromType(type) {
        switch (type) {
            case 'aws':
                return <img alt='aws' src='/assets/aws-icon.svg'/>;
            case 'cubbyhole':
                return <LockOpenIcon/>;
            case 'azure':
                return <img alt='azure' src='/assets/azure-icon.svg'/>;
            default:
                return <ListIcon/>;
        }
    }

    /**
     * Required React Component lifecycle method. Invoked once, only on the client (not on the server), immediately after the initial rendering occurs.
     *
     * @protected
     * @override
     */
    componentDidMount() {
        const {checkSession, listRequests} = this.props;
        checkSession().then(() => {
            const {listMounts, getSealStatus, vaultLookupSelf} = this.props;

            if (vaultLookupSelf.data.data.policies.includes('root')) {
                localStorageUtil.removeItem(localStorageUtil.KEY_NAMES.VAULT_TOKEN);
                this.setState({
                    showRootWarning: true
                });
            }
            getSealStatus();
            listMounts();
        });
        listRequests();
    }

    /**
     * Required React Component lifecycle method. Invoked when a component did update. This method is not called for the initial render.
     *
     * @protected
     * @override
     * @param {Object} prevProps - previous set of props.
     */
    componentDidUpdate(prevProps) {
        const {isLoggedIn, vaultLookupSelf} = this.props;
        const isRootLoggedIn = vaultLookupSelf.data && vaultLookupSelf.data.data.policies.includes('root');
        if (prevProps.isLoggedIn !== isLoggedIn || !isRootLoggedIn && localStorageUtil.getItem(localStorageUtil.KEY_NAMES.VAULT_TOKEN) === null) {
            localStorageUtil.removeItem(localStorageUtil.KEY_NAMES.VAULT_TOKEN);
            window.location.href = '/';
        }
    }

    /**
     * Required React Component lifecycle method. Returns a tree of React components that will render to HTML.
     *
     * @protected
     * @override
     * @returns {React.ReactElement}
     */
    render() {
        const {classes, logout, secretsMounts = {}, secretsRequests = [], sealStatus} = this.props;
        const isVaultSealed = sealStatus && sealStatus.sealed;
        const {accountAnchorElement, isSecretRequestsModalOpen, isUserProfileModalOpen, notificationAnchorElement, showRootWarning} = this.state;
        const rootMessage = 'You have logged in with a root token. As a security precaution, this root token will not be stored by your browser and you will need to re-authenticate after the window is closed or refreshed.';
        return <div>
            <AppBar position='static'>
                <Toolbar>
                    <Button color='inherit' component={props => <Link to='/' {...props}/>} variant='text'>
                        <img alt='logo' className='mr-1' src='/assets/vault-dark.svg'/>
                        <Typography noWrap className={classes.title} color='inherit' variant='h6'>{
                            Constants.APP_TITLE
                        }</Typography>
                    </Button>
                    <div className={classes.grow}/>
                    <Typography color={isVaultSealed ? 'secondary' : 'inherit'}>
                        Status:
                    </Typography>
                    <div className={classes.sealStatusDivider}>
                        {isVaultSealed ? <LockIcon/> : <LockOpenIcon/>}
                    </div>
                    <div className={classes.sectionDesktop}>
                        <IconButton color='inherit' onClick={() => this._openModal('isSecretRequestsModalOpen')}>
                            <Badge badgeContent='Mock data!' color='secondary'>
                                <NotificationsIcon/>
                            </Badge>
                        </IconButton>
                        <IconButton color='inherit' onClick={(event) => {
                            this.setState({
                                notificationAnchorElement: event.currentTarget
                            });
                        }}>
                            {
                                secretsRequests && secretsRequests.length > 0 ?
                                    <Badge badgeContent={secretsRequests.length} color='secondary'>
                                        <NotificationsIcon/>
                                    </Badge> :
                                    <NotificationsIcon/>
                            }
                        </IconButton>
                        <IconButton
                            aria-haspopup='true'
                            aria-owns={accountAnchorElement ? 'material-appbar' : undefined}
                            color='inherit'
                            onClick={this._toggleAccountMenu}>
                            <AccountCircle/>
                        </IconButton>
                        <Menu
                            anchorEl={accountAnchorElement}
                            open={!!accountAnchorElement}
                            onClose={this._toggleAccountMenu}>
                            <MenuItem onClick={() => this._openModal('isUserProfileModalOpen')}>
                                <img
                                    className={classes.marginRight}
                                    src='/assets/settings-icon.svg'
                                    width='20'/> Profile
                            </MenuItem>
                            <MenuItem onClick={logout}>
                                <img className={classes.marginRight} src='/assets/logout-icon.svg' width='20'/> Log Out
                            </MenuItem>
                        </Menu>
                    </div>
                </Toolbar>
            </AppBar>
            <Grid container className='mt-1' justify='center'>
                <Switch>
                    <Route exact path='/'>
                        <Card className={classes.card}>
                            <CardContent>
                                <Typography gutterBottom color='textSecondary' variant='h6'>
                                    Secrets Engines
                                </Typography>
                            </CardContent>
                            <List>{
                                (secretsMounts.data || []).map(mount => {
                                    const {description, name, type} = mount;
                                    return <ListItem
                                        button
                                        component={(props) => <Link to={`secrets/${name}`} {...props}/>}
                                        key={name}>
                                        <ListItemIcon>{
                                            this._renderIconFromType(type)
                                        }</ListItemIcon>
                                        <ListItemText primary={name} secondary={description}/>
                                        <ListItemSecondaryAction>
                                            <IconButton>
                                                <KeyboardArrowRightIcon/>
                                            </IconButton>
                                        </ListItemSecondaryAction>
                                    </ListItem>;
                                })
                            }</List>
                        </Card>
                    </Route>
                    <Route component={SecretsList} path='/secrets/:mount/:path*'/>
                    <Redirect to='/'/>
                </Switch>
            </Grid>
            <Snackbar action={
                <IconButton aria-label='Close' className={classes.close} color='inherit' onClick={this._onClose}>
                    <CloseIcon/>
                </IconButton>} anchorOrigin={{
                horizontal: 'left',
                vertical: 'bottom'
            }} autoHideDuration={12000} ContentProps={{
                classes: {message: classes.warningMessageContentWidth}
            }} message={rootMessage} open={showRootWarning} onClose={this._onClose}/>
            <SecretRequestQueueModal
                open={isSecretRequestsModalOpen}
                onClose={() => this._closeModal('isSecretRequestsModalOpen')}/>
            <UserProfileModal
                open={isUserProfileModalOpen}
                onClose={() => this._closeModal('isUserProfileModalOpen')}/>
            <NotificationsModal open={!!notificationAnchorElement} onClose={() => this.setState({
                notificationAnchorElement: null
            })}/>
            <Footer/>
        </div>;
    }
}

Main.propTypes = {
    checkSession: PropTypes.func.isRequired,
    classes: PropTypes.object.isRequired,
    getSealStatus: PropTypes.func.isRequired,
    isLoggedIn: PropTypes.bool,
    listMounts: PropTypes.func.isRequired,
    listRequests: PropTypes.func.isRequired,
    logout: PropTypes.func.isRequired,
    sealStatus: PropTypes.object.isRequired,
    secretsMounts: PropTypes.object,
    secretsRequests: PropTypes.array,
    vaultDomain: PropTypes.object.isRequired,
    vaultLookupSelf: PropTypes.object.isRequired
};

/**
 * Returns the Redux store's state that is relevant to this class as props.
 *
 * @private
 * @param {Object} state - The initial state.
 * @returns {Object}
 */
const _mapStateToProps = (state) => {
    return {
        ...state.localStorageReducer,
        ...state.sessionReducer,
        ...state.kvReducer,
        ...state.userReducer
    };
};

/**
 * Returns a map of methods used for dispatching actions to the store.
 *
 * @private
 * @param {function} dispatch Redux dispatch function.
 * @returns {Object}
 */
const _mapDispatchToProps = (dispatch) => {
    return {
        checkSession: () => {
            return new Promise((resolve, reject) => {
                dispatch(sessionAction.validateToken()).then(() => {
                    dispatch(sessionAction.setToken(localStorageUtil.getItem(localStorageUtil.KEY_NAMES.VAULT_TOKEN)));
                    resolve();
                }).catch(reject);
            });
        },
        listMounts: () => dispatch(kvAction.listMounts()),
        listRequests: () => dispatch(kvAction.listRequests()),
        getSealStatus: () => dispatch(systemAction.getSealStatus()),
        logout: () => dispatch(userAction.logout())
    };
};

/**
 * Returns custom style overrides.
 *
 * @private
 * @param {Object} theme The theme object.
 * @returns {Object}
 */
const _styles = (theme) => ({
    card: {
        width: '800px'
    },
    grow: {
        flexGrow: 1,
    },
    marginRight: {
        marginRight: theme.spacing.unit
    },
    sealStatusDivider: {
        borderRight: '0.1em solid white',
        padding: '0.5em'
    },
    sectionDesktop: {
        display: 'flex'
    },
    textCenter: {
        textAlign: 'center'
    },
    warningMessageContentWidth: {
        width: 'calc(100% - 70px)'
    }
});

export default withRouter(connect(_mapStateToProps, _mapDispatchToProps)(withStyles(_styles)(Main)));
