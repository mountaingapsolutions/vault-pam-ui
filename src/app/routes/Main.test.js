/* global it */
import {createMemoryHistory} from 'history';
import React from 'react';
import ReactDOM from 'react-dom';
import {applyMiddleware, combineReducers, createStore} from 'redux';
import {Provider} from 'react-redux';
import {Router} from 'react-router-dom';

import sessionReducer from 'app/core/reducers/sessionReducer';
import systemReducer from 'app/core/reducers/systemReducer';
import userReducer from 'app/core/reducers/userReducer';

import Main from './Main';
import reduxThunk from 'redux-thunk';

/**
 * Configures the application store by invoking Redux's createStore method.
 *
 * @private
 * @param {Object} [initialState] = The initial state.
 * @returns {Object}
 */
const _configureStore = (initialState) => {
    return createStore(
        combineReducers({
            sessionReducer,
            systemReducer,
            userReducer
        }),
        initialState,
        applyMiddleware(reduxThunk)
    );
};

it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Provider store={_configureStore()}>
        <Router history={createMemoryHistory()}>
            <Main/>
        </Router>
    </Provider>, div);
    ReactDOM.unmountComponentAtNode(div);
});
