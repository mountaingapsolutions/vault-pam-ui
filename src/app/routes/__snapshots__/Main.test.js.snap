// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <header
    className="MuiPaper-root MuiPaper-elevation4 MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <a
        aria-disabled={false}
        className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-sizeSmall MuiButton-colorInherit"
        href="/"
        onBlur={[Function]}
        onClickCapture={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={0}
      >
        <span
          className="MuiButton-label"
        >
          <img
            alt="logo"
            className="mr-1"
            src="/assets/vault-dark.svg"
          />
          <h6
            className="MuiTypography-root MuiTypography-h6 MuiTypography-colorInherit MuiTypography-noWrap"
          >
            Vault PAM UI
          </h6>
        </span>
      </a>
      <div
        className="Main-grow-2"
      />
      <span
        className="MuiTypography-root MuiTypography-overline MuiTypography-colorInherit"
      />
      <div
        className="Main-userNameDivider-6"
      />
      <p
        className="MuiTypography-root MuiTypography-body1 MuiTypography-colorInherit"
      >
        Status:
      </p>
      <div
        className="Main-sealStatusDivider-4"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 17c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm6-9h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6h1.9c0-1.71 1.39-3.1 3.1-3.1 1.71 0 3.1 1.39 3.1 3.1v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zm0 12H6V10h12v10z"
          />
        </svg>
      </div>
      <div
        className="Main-sectionDesktop-5"
      >
        <button
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 22c1.1 0 2-.9 2-2h-4c0 1.1.89 2 2 2zm6-6v-5c0-3.07-1.64-5.64-4.5-6.32V4c0-.83-.67-1.5-1.5-1.5s-1.5.67-1.5 1.5v.68C7.63 5.36 6 7.92 6 11v5l-2 2v1h16v-1l-2-2z"
              />
            </svg>
          </span>
        </button>
        <button
          aria-haspopup="true"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <img
              alt="profile"
              className=""
              onError={[Function]}
              src="//www.gravatar.com/avatar/3c133789322e4d7c8d066e9a8cf24eaf?s=24&d=404"
            />
          </span>
        </button>
      </div>
    </div>
  </header>
  <div
    className="MuiGrid-root mt-1 MuiGrid-container MuiGrid-justify-xs-center"
  >
    <div
      className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded MuiCard-root Main-card-1"
    >
      <div
        className="MuiCardContent-root"
      >
        <h6
          className="MuiTypography-root MuiTypography-h6 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
        >
          Secrets Engines
        </h6>
      </div>
      <ul
        className="MuiList-root MuiList-padding"
      />
    </div>
  </div>
  <div
    className="Footer-footerRootContainer-226"
  >
    <header
      className="MuiPaper-root MuiPaper-elevation4 MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary mui-fixed Footer-footer-225"
    >
      <div
        className="Footer-footerTextContainer-227"
      >
        <a
          className="MuiTypography-root MuiTypography-caption MuiTypography-displayInline Footer-text-230"
          href="/rest/api"
          title="API"
        >
          API
        </a>
        <span
          className="MuiTypography-root MuiTypography-caption MuiTypography-displayInline Footer-textSeparator-229"
        >
          |
        </span>
        <span
          className="MuiTypography-root MuiTypography-caption MuiTypography-displayInline Footer-textActiveDomain-228"
        >
          
        </span>
        <span
          className="MuiTypography-root MuiTypography-caption MuiTypography-displayInline Footer-textSeparator-229"
        >
          |
        </span>
        <span
          className="MuiTypography-root MuiTypography-caption MuiTypography-displayInline Footer-text-230"
        >
          Vault 
        </span>
      </div>
    </header>
  </div>
</div>
`;
